netcreateorg/netcreate-2018
# Feature: Define Prompts and Colors via Template

You can now define both prompts and colors by editing a template file.

You can:
* Add/edit/remove node types
* Add/edit/remove edge types
* Set color for each node type.

The template is currently at `build/app/assets/templates/alexander.json`.  

The file is in JSON format, so you'll need to be careful with commas and brackets when editing.  Colors should update as you edit the json file.

# TO TEST

1. Check out `dev-bl/template-system` branch
2. Start the app
3. The Alexander data should be loaded, with nodes color coded.
4. Try changing the colors defined in the `build/app/assets/templates/alexander.json` file.
5. Try adding a node option type.  Create a node to see if it works.
6. Try removing a node option type.  View an existing node to see how the label is handled.
7.  Try changing the label for a node option type.  Nodes with the old label will be displayed in black (non-matching node).  NOTE: Using label rather than id means we have no way of making the mapping from the old label to the new label.
7.  Try adding and removing an edge option type.  View the results to make sure it works.


# Caveats

* **Undefined Types** -- Right now node colors are defined by matching the text of the selected "Type" (`node.attributes["Node_Type"]`).  The advantage of using this method at this point is backward compatibility with previously created data.  However, if you should change the label prompts, then the color lookup table will fail.  Any unmatched labels are drawn as gray.  We need to have a conversation about data structures and how you want to save data.  Specifically, whether we should be saving node types as an id or a label field.


# Still To Do

* **Selectable Template Files** -- Right now while you can choose from any number of template files, you need to set the active one via modifying the load call in `nc-logic.jsx`.  In the future, we could make it easier to select an alternative template.

* **YAML** -- YAML might be slightly easier to edit.  If we don't end up building a UI, we may convert the template file to a YAML.  We just don't currently have YAML file-loading in the system, so to save time/money we used the existing JSON file loader.

* **Unused Parameters** -- Right now only the `nodePrompt.type` and `edgePrompt.type` are being used.  The other prompts and labels are there as placeholders for the future for when we might make nodes and edges fields completely definable from the template file.
